### ISSUE TITLE
🗺️ Middleware

### ISSUE THREAD

**Issue by ryanflorence** [remix-run/react-router#12695]
- RFC: https://github.com/remix-run/remix/discussions/7642
- Supersedes https://github.com/remix-run/react-router/issues/9566

**Comment by airtonix** [comment]
what about ssr? All of the above looks like browser only routing.

Surely react router 7 isn't being marketed as the successor "right now" to remix without this?

**Comment by timdorr** [comment]
@airtonix You're looking at old comments from before RR7 was released.

**Comment by airtonix** [comment]
> [@airtonix](https://github.com/airtonix) You're looking at old comments from before RR7 was released.

Sure but where's the middleware documentation for react router 7? 

All i see is `unstable` mentions in the "releases" marketing page...

![Image](https://github.com/user-attachments/assets/88a39533-f28b-4d0e-936c-b017c49e12f3)

Where do I find the documentation for React Router 7 Stable? 

And just to be clear, I'm not really interested in client side browser only middleware.

**Comment by sergiodxa** [comment]
@airtonix there's no docs yet because it's still unstable, the release change log is the only official things, I have been using them and wrote about how to use it here https://sergiodxa.com/tutorials/use-middleware-in-react-router

**Comment by brophdawg11** [comment]
Re-opening to reflect this on the updated [roadmap](https://github.com/orgs/remix-run/projects/5) for the new [governance](https://github.com/remix-run/react-router/blob/main/GOVERNANCE.md) process

We have also since added some docs on the unstable middleware APIs: https://reactrouter.com/how-to/middleware

**Comment by thomasgadsden** [comment]
We've been trying to get `unstable_clientMiddleware` to run using react-router@7.7.0 with `createBrowserRouter` (manual routing) in a Vite project. Despite all our efforts, the client middleware never runs. Does `unstable_clientMiddleware` only work in file-based routing, or are there other steps we need to take to use it with manual routing?

**Comment by brophdawg11** [comment]
yeah `clientMiddleware` is a framework-mode thing because that's where the distinction between server and client comes in (`loader`/`clientLoader`, `middleware`/`clientMiddleware`).  For data routers, everything is "client" so there's no distinction.  You can just put `unstable_middleware: [...]` on your route object

**Comment by thomasgadsden** [comment]
@brophdawg11 thanks for the clarification on this. I changed our code to use `unstable_middleware` but it still wasn’t working. Eventually, I found some other issues and discussions here and realised that we weren’t defining a `loader` function on our route-objects – once I added this, everything worked as expected 👍

**Comment by brophdawg11** [comment]
Released as stable in 7.9.0
